[scroll-animations] Have animation-range parsing handle sequences

nmoucht opened the pull request:
#### 83dcd4537698ccc0d6bbbfc6b636a95dec36403d
<pre>
[scroll-animations] Have animation-range parsing handle sequences
<a href="https://bugs.webkit.org/show_bug.cgi?id=282004">https://bugs.webkit.org/show_bug.cgi?id=282004</a>
<a href="https://rdar.apple.com/138507260">rdar://138507260</a>

Reviewed by Tim Nguyen.

Handle comma seperated sequences of animation ranges.

* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/parsing/animation-range-end-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/parsing/animation-range-end-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/parsing/animation-range-shorthand-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/parsing/animation-range-shorthand.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/parsing/animation-range-start-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/parsing/animation-range-start-valid-expected.txt:
* Source/WebCore/animation/TimelineRange.cpp:
(WebCore::SingleTimelineRange::parse):
* Source/WebCore/css/ShorthandSerializer.cpp:
(WebCore::ShorthandSerializer::serializeSingleAnimationRange const):
(WebCore::ShorthandSerializer::serializeAnimationRange const):
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::CSSPropertyParser::consumeAnimationRangeShorthand):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Timeline.cpp:
(WebCore::CSSPropertyParserHelpers::consumeAnimationRange):
(WebCore::CSSPropertyParserHelpers::consumeAnimationRangeStart):
(WebCore::CSSPropertyParserHelpers::consumeAnimationRangeEnd):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Timeline.h:

Canonical link: <a href="https://commits.webkit.org/285656@main">https://commits.webkit.org/285656@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/36d3e983b025a6d87a97bf08ac85fac4c844f812

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/73342 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/52771 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/26150 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/77577 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/24580 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/75457 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/60577 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/554 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/57614 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16080 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/76409 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/47656 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/63106 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/38032 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/44296 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/20587 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/22909 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/66149 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/20940 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/79224 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/653 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/176 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/66043 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/799 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/63117 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/65320 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/9167 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/7338 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11304 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/621 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/3358 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/651 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/635 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/653 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/ShorthandSerializer.cpp:
```diff
@@ -1343,16 +1343,23 @@ String ShorthandSerializer::serializeSingleAnimationRange(CSSValue& value, Singl
-
-    auto serializedStart = serializeSingleAnimationRange(start, SingleTimelineRange::Type::Start);
-    auto serializedEnd = serializeSingleAnimationRange(longhandValue(1), SingleTimelineRange::Type::End, startID);
+    StringBuilder serializedString;
```
we usually name this variable `builder`

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/css/ShorthandSerializer.cpp:
```diff
@@ -1343,16 +1343,23 @@ String ShorthandSerializer::serializeSingleAnimationRange(CSSValue& value, Singl
+            auto serializedStart = serializeSingleAnimationRange(*start, SingleTimelineRange::Type::Start);
+            auto serializedEnd = serializeSingleAnimationRange(*endList->item(i), SingleTimelineRange::Type::End, startID);
+            serializedString.append(serializedEnd.isEmpty() ? serializedStart : makeString(serializedStart, ' ', serializedEnd), (i < startList->size() - 1) ? ", "_s : emptyString());
+        }
```
```suggestion
            serializedString.append(
                serializedEnd.isEmpty() ? serializedStart : makeString(serializedStart, ' ', serializedEnd),
                i == startList->size() - 1 ? ""_s : ", "_s
            );
        }
```

Can we format like this? It would be even better if you can use `interleave()` instead of this loop.

nmoucht replied:
Fixed. I don't think we can use interleave on a `CSSValueList`.

nt1m commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -2775,38 +2775,48 @@ bool CSSPropertyParser::consumeWhiteSpaceShorthand(bool important)
 
-    addProperty(CSSPropertyAnimationRangeStart, CSSPropertyAnimationRange, WTFMove(start), important);
-    addProperty(CSSPropertyAnimationRangeEnd, CSSPropertyAnimationRange, WTFMove(end), important);
+        RefPtr<CSSValue> end = nullptr;
```
```suggestion
        RefPtr<CSSValue> end;
```

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -2775,38 +2775,48 @@ bool CSSPropertyParser::consumeWhiteSpaceShorthand(bool important)
+        }
+        startList.append(start.releaseNonNull());
+        endList.append(end.releaseNonNull());
+
```
nit: remove this line

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Timeline.cpp:
```diff
@@ -180,12 +180,16 @@ static RefPtr<CSSValue> consumeAnimationRange(CSSParserTokenRange& range, const
 RefPtr<CSSValue> consumeAnimationRangeEnd(CSSParserTokenRange& range, const CSSParserContext& context)
 {
-    return consumeAnimationRange(range, context, SingleTimelineRange::Type::End);
+    return consumeCommaSeparatedListWithSingleValueOptimization(range, [context](CSSParserTokenRange& range) -> RefPtr<CSSValue> {
```
```suggestion
    return consumeCommaSeparatedListWithSingleValueOptimization(range, [&](CSSParserTokenRange& range) -> RefPtr<CSSValue> {
```

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Timeline.cpp:
```diff
@@ -180,12 +180,16 @@ static RefPtr<CSSValue> consumeAnimationRange(CSSParserTokenRange& range, const
 RefPtr<CSSValue> consumeAnimationRangeStart(CSSParserTokenRange& range, const CSSParserContext& context)
 {
-    return consumeAnimationRange(range, context, SingleTimelineRange::Type::Start);
+    return consumeCommaSeparatedListWithSingleValueOptimization(range, [context](CSSParserTokenRange& range) -> RefPtr<CSSValue> {
```
```suggestion
    return consumeCommaSeparatedListWithSingleValueOptimization(range, [&](CSSParserTokenRange& range) -> RefPtr<CSSValue> {
```

nmoucht replied:
Fixed.

nmoucht committed:
[scroll-animations] Have animation-range parsing handle sequences
https://bugs.webkit.org/show_bug.cgi?id=282004
rdar://138507260

Reviewed by Tim Nguyen.

Handle comma seperated sequences of animation ranges.

* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/parsing/animation-range-end-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/parsing/animation-range-end-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/parsing/animation-range-shorthand-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/parsing/animation-range-shorthand.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/parsing/animation-range-start-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/parsing/animation-range-start-valid-expected.txt:
* Source/WebCore/animation/TimelineRange.cpp:
(WebCore::SingleTimelineRange::parse):
* Source/WebCore/css/ShorthandSerializer.cpp:
(WebCore::ShorthandSerializer::serializeSingleAnimationRange const):
(WebCore::ShorthandSerializer::serializeAnimationRange const):
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::CSSPropertyParser::consumeAnimationRangeShorthand):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Timeline.cpp:
(WebCore::CSSPropertyParserHelpers::consumeAnimationRange):
(WebCore::CSSPropertyParserHelpers::consumeAnimationRangeStart):
(WebCore::CSSPropertyParserHelpers::consumeAnimationRangeEnd):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Timeline.h:

Canonical link: https://commits.webkit.org/285656@main
